aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -174,6 +174,85 @@ Singleton lifetime services are created the first time they're requested (or whe
 
+## Service registration methods
+
+Each service registration extension method offers overloads that are useful in specific scenarios.
+
+**Option A: 2-level list**
+
+* In the following scenarios, the service container creates the service instance and disposes types that it creates automatically:
+
+  * Concrete type implements an interface &ndash; Useful when unit or integration tests mock objects for testing.
+
+    **Signature**: `Add{LIFETIME}<{INTERFACE}, {TYPE}>()`  
+    **Example**: `services.AddScoped<IMyDependency, MyDependency>();`
+
+    **Signature**: `Add{LIFETIME}<{INTERFACE}>(sp => new {TYPE})`  
+    **Examples**:  
+    `services.AddScoped<IMyDependency>(sp => new MyDependency());`  
+    `services.AddScoped<IMyDependency>(sp => new MyDependency("A string!"));`
+
+  * Concrete type is registered directly in the container:
+
+    **Signature**: `Add{LIFETIME}<{TYPE}>()`  
+    **Example**: `services.AddScoped<MyDependency>();`
+
+* In the following scenarios, the service container doesn't create the service instance, so user code must dispose of the instances created by the type:
+
+  * Concrete type implements an interface &ndash; Useful when unit or integration tests mock objects for testing.
+
+    **Signature**: `Add{LIFETIME}<{INTERFACE}>(new {TYPE})`  
+    **Examples**:  
+    `services.AddScoped<IMyDependency>(new MyDependency());`  
+    `services.AddScoped<IMyDependency>(new MyDependency("A string!"));`
+
+  * Concrete type is registered directly in the container:
+
+    **Signature**: `Add{LIFETIME}(new {TYPE})`  
+    **Examples**:  
+    `services.AddScoped(new MyDependency());`  
+    `services.AddScoped(new MyDependency("A string!"));`
+
+**Option B: Table**
+
+| Method | Automatic<br>object<br>disposal | Mocks for<br>testing | Pass args |
+| ------ | :-----------------------------: | :---------------------------------: | :---------------------------------: |
+| `Add{LIFETIME}<{INTERFACE}, {TYPE}>()`<br>Example:<br>`services.AddScoped<IMyDep, MyDep>();` | Yes | Yes | No |
+| `Add{LIFETIME}<{INTERFACE}>(sp => new {TYPE})`<br>Examples:<br>`services.AddScoped<IMyDep>(sp => new MyDep());`<br>`services.AddScoped<IMyDep>(sp => new MyDep("A string!"));` | Yes | Yes | Yes |
+| `Add{LIFETIME}<{TYPE}>()`<br>Example:<br>`services.AddScoped<MyDep>();` | Yes | No | No |
+| `Add{LIFETIME}<{INTERFACE}>(new {TYPE})`<br>Examples:<br>`services.AddScoped<IMyDep>(new MyDep());`<br>`services.AddScoped<IMyDep>(new MyDep("A string!"));` | No | Yes | Yes |
+| `Add{LIFETIME}(new {TYPE})`<br>Examples:<br>`services.AddScoped(new MyDep());`<br>`services.AddScoped(new MyDep("A string!"));` | No | No | Yes |
+
+For more information on type disposal, see the [Disposal of services](#disposal-of-services) section. For more information on mocking types for testing, see <xref:test/integration-tests#inject-mock-services>.
+
+`TryAdd{LIFETIME}` methods only register the service if there isn't already a registered instance of the service.
```

Comment:
```suggestion
`TryAdd{LIFETIME}` methods only register the service if there isn't already an implementation registered for the service.
```

I want to try and make it clear that TryAdd only checks the `{INTERFACE}` value. For example:

```csharp
services.TryAddSingleton<IFrob, DefaultFrob>();
services.TryAddSingleton<IFrob, AlternateFrob>(); // No-op, even though the implementation type is different
```